README.md
# Important

This repository was moved to a monorepo at github.com/twilio-labs/serverless-toolkit.

For the old version of the code please check out the `v4` branch.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [4.1.1](https://github.com/twilio-labs/serverless-api/compare/v4.1.0...v4.1.1) (2021-03-31)

### Bug Fixes

* **pagination:** `got` changed behaviour around setting a `prefixUrl` to `undefined` leading to our pagination requests 404ing because the request URL looked like `https://serverless.twilio.com/v1/https://serverless.twilio.com/v1/PATH`. Setting `prefixUrl` to `""` fixes this.

## [4.1.0](https://github.com/twilio-labs/serverless-api/compare/v4.0.3...v4.1.0) (2021-02-17)

### Features

* **runtime:** adds runtime parameter to config ([c92ebcb](https://github.com/twilio-labs/serverless-api/commit/c92ebcb1ac2e226247080e9a9bf1ee814595461e))

### [4.0.3](https://github.com/twilio-labs/serverless-api/compare/v4.0.2...v4.0.3) (2020-10-29)


### Bug Fixes

* **logs:** change caching of seen log entries ([#58](https://github.com/twilio-labs/serverless-api/issues/58)) ([7ebc5b3](https://github.com/twilio-labs/serverless-api/commit/7ebc5b3bc08c1ce23c27d9d4f138b5e62954802e)), closes [twilio-labs/serverless-toolkit#192](https://github.com/twilio-labs/serverless-toolkit/issues/192)

### [4.0.2](https://github.com/twilio-labs/serverless-api/compare/v4.0.1...v4.0.2) (2020-08-27)


### Bug Fixes

* add 429 handling from twilio-labs/serverless-toolkit[#178](https://github.com/twilio-labs/serverless-api/issues/178) ([449fda2](https://github.com/twilio-labs/serverless-api/commit/449fda27dedaa6331ad73ee45bfd477c3c2149bf))

### [4.0.1](https://github.com/twilio-labs/serverless-api/compare/v4.0.0...v4.0.1) (2020-07-28)


### Bug Fixes

* upgrade type-fest from 0.15.1 to 0.16.0 ([bfbcdc9](https://github.com/twilio-labs/serverless-api/commit/bfbcdc9b281ea94477f8f2c88c0a9fc7a36a9214))

## [4.0.0](https://github.com/twilio-labs/serverless-api/compare/v3.0.0...v4.0.0) (2020-07-06)


### Features

* add tests for asset error ([#47](https://github.com/twilio-labs/serverless-api/issues/47)) ([581e4ce](https://github.com/twilio-labs/serverless-api/commit/581e4ce7bde33c06e394819f749dc09fee1fdd19))
* **client:** adds username and password support ([#51](https://github.com/twilio-labs/serverless-api/issues/51)) ([80ffe00](https://github.com/twilio-labs/serverless-api/commit/80ffe008a2db0b2d2aa294328a35e0d3c359a19a)), closes [#41](https://github.com/twilio-labs/serverless-api/issues/41)
* **functions:** list function versions & download content ([#52](https://github.com/twilio-labs/serverless-api/issues/52)) ([045a90a](https://github.com/twilio-labs/serverless-api/commit/045a90a536e850f84388cc5dcc06e6f2c0c7b9b5))


### Bug Fixes

* add file name on failing asset upload error ([#47](https://github.com/twilio-labs/serverless-api/issues/47)) ([13fedb2](https://github.com/twilio-labs/serverless-api/commit/13fedb2544acda4d38928fde035eb15783432485))
* **activate:** activating with environment variables ([#49](https://github.com/twilio-labs/serverless-api/issues/49)) ([62c2af6](https://github.com/twilio-labs/serverless-api/commit/62c2af664e56197034f263a309fd73b50d7f38a9))
* **deploy:** stop creating environments with domain suffix 'undefined' ([#50](https://github.com/twilio-labs/serverless-api/issues/50)) ([f5dde4c](https://github.com/twilio-labs/serverless-api/commit/f5dde4c55997d6214f8b930a2dc09a0f09c01e66))

### [3.1.1](https://github.com/twilio-labs/serverless-api/compare/v3.1.0...v3.1.1) (2020-06-25)


### Bug Fixes

* **deploy:** stop creating environments with domain suffix 'undefined' ([#50](https://github.com/twilio-labs/serverless-api/issues/50)) ([f5dde4c](https://github.com/twilio-labs/serverless-api/commit/f5dde4c55997d6214f8b930a2dc09a0f09c01e66))

## [3.1.0](https://github.com/twilio-labs/serverless-api/compare/v3.0.0...v3.1.0) (2020-06-16)


### Features

* add tests for asset error ([#47](https://github.com/twilio-labs/serverless-api/issues/47)) ([581e4ce](https://github.com/twilio-labs/serverless-api/commit/581e4ce7bde33c06e394819f749dc09fee1fdd19))


### Bug Fixes

* **activate:** activating with environment variables ([#49](https://github.com/twilio-labs/serverless-api/issues/49)) ([62c2af6](https://github.com/twilio-labs/serverless-api/commit/62c2af664e56197034f263a309fd73b50d7f38a9))
* add file name on failing asset upload error ([#47](https://github.com/twilio-labs/serverless-api/issues/47)) ([13fedb2](https://github.com/twilio-labs/serverless-api/commit/13fedb2544acda4d38928fde035eb15783432485))

## [3.0.0](https://github.com/twilio-labs/serverless-api/compare/v2.1.0...v3.0.0) (2020-05-21)


### ⚠ BREAKING CHANGES

* the function signature of the individual functions changed. If you only use the methods on the client, nothing should have changed.

### Features

* introduce rate limiting ([#42](https://github.com/twilio-labs/serverless-api/issues/42)) ([6dc1bc4](https://github.com/twilio-labs/serverless-api/commit/6dc1bc43e7516d816a976ea190e7fc4b35a12f3f)), closes [#37](https://github.com/twilio-labs/serverless-api/issues/37)

## [2.1.0](https://github.com/twilio-labs/serverless-api/compare/v1.1.0...v2.1.0) (2020-05-18)


### Features

* add region support ([#39](https://github.com/twilio-labs/serverless-api/issues/39)) ([afe4149](https://github.com/twilio-labs/serverless-api/commit/afe414915932074f64148d94a778d6be8ea22b16))

src/client.ts
/** @module @twilio-labs/serverless-api */

import debug from 'debug';
import events from 'events';
import {
  HTTPAlias,
  Options,
  OptionsOfJSONResponseBody,
  OptionsOfTextResponseBody,
  Response,
} from 'got/dist/source';
import pLimit, { Limit } from 'p-limit';
import { deprecate } from 'util';
import { getOrCreateAssetResources, uploadAsset } from './api/assets';
import {
  activateBuild,
  getBuild,
  listBuilds,
  triggerBuild,
  waitForSuccessfulBuild,
} from './api/builds';
import { getDependencies } from './api/dependencies';
import {
  createEnvironmentFromSuffix,
  createEnvironmentIfNotExists,
  getEnvironment,
  getEnvironmentFromSuffix,
  isEnvironmentSid,
  listEnvironments,
} from './api/environments';
import {
  getOrCreateFunctionResources,
  isFunctionSid,
  listFunctionResources,
  uploadFunction,
} from './api/functions';
import { listOnePageLogResources } from './api/logs';
import { createService, findServiceSid, listServices } from './api/services';
import { getApiUrl } from './api/utils/api-client';
import { CONCURRENCY, RETRY_LIMIT } from './api/utils/http_config';
import {
  listVariablesForEnvironment,
  setEnvironmentVariables,
} from './api/variables';
import got from './got';
import { LogsStream } from './streams/logs';
import {
  ActivateConfig,
  ActivateResult,
  BuildResource,
  ClientConfig,
  DeployLocalProjectConfig,
  DeployProjectConfig,
  DeployResult,
  GotClient,
  ListConfig,
  ListResult,
  LogApiResource,
  LogsConfig,
} from './types';
import { DeployStatus } from './types/consts';
import { ClientApiError, convertApiErrorsAndThrow } from './utils/error';
import { getListOfFunctionsAndAssets, SearchConfig } from './utils/fs';

const log = debug('twilio-serverless-api:client');

export function createGotClient(config: ClientConfig): GotClient {
  let username, password;
  if ('accountSid' in config) {
    username = config.accountSid;
    password = config.authToken;
    deprecate(() => {},
    '`accountSid` and `authToken` client config is deprecated, please use `username` and `password` instead.')();
  } else {
    username = config.username;
    password = config.password;
  }
  const client = got.extend({
    prefixUrl: getApiUrl(config),
    responseType: 'json',
    username: username,
    password: password,
    headers: {
      'User-Agent': 'twilio-serverless-api',
    },
  }) as GotClient;
  return client;
}

export class TwilioServerlessApiClient extends events.EventEmitter {
  /**
   * Contains the client config used to do API requests
   *
   * @private
   * @type {ClientConfig}
   * @memberof TwilioServerlessApiClient
   */
  private config: ClientConfig;

  /**
   * The `got` client that is used to make all the API requests
   *
   * @private
   * @type {GotClient}
   * @memberof TwilioServerlessApiClient
   */
  private client: GotClient;

  /**
   *
   * @private
   * @type {Limit}
   * @memberof TwilioServerlessApiClient
   */
  private limit: Limit;

  constructor(config: ClientConfig) {
    debug.enable(process.env.DEBUG || '');
    super();
    this.config = config;
    this.client = createGotClient(config);
    this.limit = pLimit(config.concurrency || CONCURRENCY);
  }

  /**
   * Returns the internally used GotClient instance used to make API requests
   * @returns {GotClient} A client instance with set-up credentials
   */
  getClient(): GotClient {
    debug.enable(process.env.DEBUG || '');
    return this.client;
  }

  /**
   * Returns an object containing lists of services, environments, variables
   * functions or assets, depending on which have beeen requested in `listConfig`
   * @param  {ListConfig} listConfig Specifies info around which things should be listed
   * @returns Promise<ListResult> Object containing the different lists.
   */
  async list(listConfig: ListConfig): Promise<ListResult> {
    try {
      let {
        types,
        serviceSid,
        serviceName: serviceName,
        environment: environmentSid,
      } = listConfig;

      if (
        types === 'services' ||
        (types.length === 1 && types[0] === 'services')
      ) {
        const services = await listServices(this);
        return { services };
      }

      if (
        typeof serviceSid === 'undefined' &&
        typeof serviceName !== 'undefined'
      ) {
        serviceSid = await findServiceSid(serviceName, this);
      }

      if (typeof serviceSid === 'undefined') {
        throw new Error('Missing service SID argument');
      }

      const result: ListResult = {};

      let currentBuildSidForEnv: string | undefined;
      let currentBuild: BuildResource | undefined;
      for (const type of types) {
        try {
          if (type === 'environments') {
            result.environments = await listEnvironments(serviceSid, this);
          }

          if (type === 'builds') {
            result.builds = await listBuilds(serviceSid, this);
          }

          if (typeof environmentSid === 'string') {
            if (!isEnvironmentSid(environmentSid)) {
              const environment = await getEnvironmentFromSuffix(
                environmentSid,
                serviceSid,
                this
              );
              environmentSid = environment.sid;
              currentBuildSidForEnv = environment.build_sid;
            } else if (!currentBuildSidForEnv) {
              const environment = await getEnvironment(
                environmentSid,
                serviceSid,
                this
              );
              currentBuildSidForEnv = environment.build_sid;
            }

            if (type === 'functions' || type === 'assets') {
              if (!currentBuild) {
                currentBuild = await getBuild(
                  currentBuildSidForEnv,
                  serviceSid,
                  this
                );
              }

              if (type === 'functions') {
                result.functions = {
                  environmentSid,
                  entries: currentBuild.function_versions,
                };
              } else if (type === 'assets') {
                result.assets = {
                  environmentSid,
                  entries: currentBuild.asset_versions,
                };
              }
            }

            if (type === 'variables') {
              result.variables = {
                entries: await listVariablesForEnvironment(
                  environmentSid,
                  serviceSid,
                  this
                ),
                environmentSid,
              };
            }
          }
        } catch (err) {
          log(new ClientApiError(err));
        }
      }

      return result;
    } catch (err) {
      convertApiErrorsAndThrow(err);
    }
  }

  async getLogsStream(logsConfig: LogsConfig): Promise<LogsStream> {
    try {
      let { serviceSid, environment, filterByFunction } = logsConfig;
      if (!isEnvironmentSid(environment)) {
        const environmentResource = await getEnvironmentFromSuffix(
          environment,
          serviceSid,
          this
        );
        environment = environmentResource.sid;
      }

      if (filterByFunction && !isFunctionSid(filterByFunction)) {
        const availableFunctions = await listFunctionResources(
          serviceSid,
          this
        );
        const foundFunction = availableFunctions.find(
          (fn) => fn.friendly_name === filterByFunction
        );
        if (!foundFunction) {
          throw new Error('Invalid Function Name or SID');
        }
        filterByFunction = foundFunction.sid;
      }
      const logsStream = new LogsStream(
        environment,
        serviceSid,
        this,
        logsConfig
      );

      return logsStream;
    } catch (err) {
      convertApiErrorsAndThrow(err);
    }
  }

  async getLogs(logsConfig: LogsConfig): Promise<LogApiResource[]> {
    try {
      let { serviceSid, environment, filterByFunction } = logsConfig;
      if (!isEnvironmentSid(environment)) {
        const environmentResource = await getEnvironmentFromSuffix(
          environment,
          serviceSid,
          this
        );
        environment = environmentResource.sid;
      }

      if (filterByFunction && !isFunctionSid(filterByFunction)) {
        const availableFunctions = await listFunctionResources(
          serviceSid,
          this
        );
        const foundFunction = availableFunctions.find(
          (fn) => fn.friendly_name === filterByFunction
        );
        if (!foundFunction) {
          throw new Error('Invalid Function Name or SID');
        }
        filterByFunction = foundFunction.sid;
      }

      return listOnePageLogResources(environment, serviceSid, this, {
        pageSize: 50,
        functionSid: filterByFunction,
      });
    } catch (err) {
      convertApiErrorsAndThrow(err);
    }
  }

  /**
   * "Activates" a build by taking a specified build SID or a "source environment"
   * and activating the same build in the specified `environment`.
   *
   * Can optionally create the new environment when called with `activateConfig.createEnvironment`
   * @param  {ActivateConfig} activateConfig Config to specify which build to activate in which environment
   * @returns Promise<ActivateResult> Object containing meta information around deployment
   */
  async activateBuild(activateConfig: ActivateConfig): Promise<ActivateResult> {
    try {
      let {
        buildSid,
        targetEnvironment,
        serviceSid,
        sourceEnvironment,
        env,
      } = activateConfig;

      if (!buildSid && !sourceEnvironment) {
        const error = new Error(
          'You need to specify either a build SID or source environment to activate'
        );
        error.name = 'activate-missing-source';
        throw error;
      }

      if (!isEnvironmentSid(targetEnvironment)) {
        try {
          const environment = await getEnvironmentFromSuffix(
            targetEnvironment,
            serviceSid,
            this
          );
          targetEnvironment = environment.sid;
        } catch (err) {
          if (activateConfig.force || activateConfig.createEnvironment) {
            const environment = await createEnvironmentFromSuffix(
              targetEnvironment,
              serviceSid,
              this
            );
            targetEnvironment = environment.sid;
          } else {
            throw err;
          }
        }
      }

      if (!buildSid && sourceEnvironment) {
        let currentEnv;
        if (!isEnvironmentSid(sourceEnvironment)) {
          currentEnv = await getEnvironmentFromSuffix(
            sourceEnvironment,
            serviceSid,
            this
          );
        } else {
          currentEnv = await getEnvironment(
            sourceEnvironment,
            serviceSid,
            this
          );
        }
        buildSid = currentEnv.build_sid;
      }

      if (!buildSid) {
        throw new Error('Could not determine build SID');
      }

      this.emit('status-update', {
        status: DeployStatus.SETTING_VARIABLES,
        message: 'Setting environment variables',
      });
      await setEnvironmentVariables(env, targetEnvironment, serviceSid, this);

      const { domain_name } = await getEnvironment(
        targetEnvironment,
        serviceSid,
        this
      );
      await activateBuild(buildSid, targetEnvironment, serviceSid, this);

      return {
        serviceSid,
        buildSid,
        environmentSid: targetEnvironment,
        domain: domain_name,
      };
    } catch (err) {
      convertApiErrorsAndThrow(err);
    }
  }

  /**
   * Deploys a set of functions, assets, variables and dependencies specified
   * in `deployConfig`. Functions & assets can either be paths to the local
   * filesystem or `Buffer` instances allowing you to dynamically upload
   * even without a file system.
   *
   * Unless a `deployConfig. serviceSid` is specified, it will try to create one. If a service
   * with the name `deployConfig.serviceName` already exists, it will throw
   * an error. You can make it use the existing service by setting `overrideExistingService` to
   * true.
   *
   * Updates to the deployment will be emitted as events to `status-update`. Example:
   *
   * ```js
   * client.on('status-update', ({ status, message }) => {
   *  console.log('[%s]: %s', status, message);
   * })
   * ```
   * @param  {DeployProjectConfig} deployConfig Config containing all details for deployment
   * @returns Promise<DeployResult> Object containing meta information around deployment
   */
  async deployProject(
    deployConfig: DeployProjectConfig
  ): Promise<DeployResult> {
    try {
      const config = {
        ...this.config,
        ...deployConfig,
      };

      const { functions, assets, runtime } = config;

      let serviceSid = config.serviceSid;
      if (!serviceSid) {
        this.emit('status-update', {
          status: DeployStatus.CREATING_SERVICE,
          message: 'Creating Service',
        });
        try {
          serviceSid = await createService(config.serviceName, this);
        } catch (err) {
          const alternativeServiceSid = await findServiceSid(
            config.serviceName,
            this
          );
          if (!alternativeServiceSid) {
            throw err;
          }
          if (config.overrideExistingService || config.force) {
            serviceSid = alternativeServiceSid;
          } else {
            const error = new Error(
              `Service with name "${config.serviceName}" already exists with SID "${alternativeServiceSid}".`
            );
            error.name = 'conflicting-servicename';
            Object.defineProperty(error, 'serviceSid', {
              value: alternativeServiceSid,
            });
            Object.defineProperty(error, 'serviceName', {
              value: config.serviceName,
            });
            throw error;
          }
        }
      }

      this.emit('status-update', {
        status: DeployStatus.CONFIGURING_ENVIRONMENT,
        message: `Configuring ${
          config.functionsEnv.length === 0 ? 'bare' : `"${config.functionsEnv}"`
        } environment`,
      });
      const environment = await createEnvironmentIfNotExists(
        config.functionsEnv,
        serviceSid,
        this
      );
      const { sid: environmentSid, domain_name: domain } = environment;

      //
      // Functions
      //

      this.emit('status-update', {
        status: DeployStatus.CREATING_FUNCTIONS,
        message: `Creating ${functions.length} Functions`,
      });
      const functionResources = await getOrCreateFunctionResources(
        functions,
        serviceSid,
        this
      );

      this.emit('status-update', {
        status: DeployStatus.UPLOADING_FUNCTIONS,
        message: `Uploading ${functions.length} Functions`,
      });
      const functionVersions = await Promise.all(
        functionResources.map((fn) => {
          return uploadFunction(fn, serviceSid as string, this, this.config);
        })
      );

      //
      // Assets
      //

      this.emit('status-update', {
        status: DeployStatus.CREATING_ASSETS,
        message: `Creating ${assets.length} Assets`,
      });
      const assetResources = await getOrCreateAssetResources(
        assets,
        serviceSid,
        this
      );

      this.emit('status-update', {
        status: DeployStatus.UPLOADING_ASSETS,
        message: `Uploading ${assets.length} Assets`,
      });
      const assetVersions = await Promise.all(
        assetResources.map((asset) => {
          return uploadAsset(asset, serviceSid as string, this, this.config);
        })
      );

      this.emit('status-update', {
        status: DeployStatus.BUILDING,
        message: 'Waiting for deployment.',
      });
      const dependencies = getDependencies(config.pkgJson);
      const build = await triggerBuild(
        { functionVersions, dependencies, assetVersions, runtime },
        serviceSid,
        this
      );
      await waitForSuccessfulBuild(build.sid, serviceSid, this, this);

      this.emit('status-update', {
        status: DeployStatus.SETTING_VARIABLES,
        message: 'Setting environment variables',
      });
      await setEnvironmentVariables(
        config.env,
        environmentSid,
        serviceSid,
        this
      );

      this.emit('status-update', {
        status: DeployStatus.ACTIVATING_DEPLOYMENT,
        message: 'Activating deployment',
      });
      await activateBuild(build.sid, environmentSid, serviceSid, this);

      this.emit('status', {
        status: DeployStatus.DONE,
        message: 'Project successfully deployed',
      });
      return {
        serviceSid,
        environmentSid,
        buildSid: build.sid,
        domain,
        functionResources,
        assetResources,
        runtime: build.runtime,
      };
    } catch (err) {
      convertApiErrorsAndThrow(err);
    }
  }

  /**
   * Deploys a local project by reading existing functions and assets
   * from `deployConfig.cwd` and calling `this.deployProject` with it.
   *
   * Functions have to be placed in a `functions` or `src` directory to be found.
   * Assets have to be placed into an `assets` or `static` directory.
   *
   * Nested folder structures will result in nested routes.
   * @param  {DeployLocalProjectConfig} deployConfig
   * @returns Promise<DeployResult> Object containing meta information around deployment
   */
  async deployLocalProject(
    deployConfig: DeployLocalProjectConfig
  ): Promise<DeployResult> {
    try {
      this.emit('status-update', {
        status: DeployStatus.READING_FILESYSTEM,
        message: 'Gathering Functions and Assets to deploy',
      });

      log('Deploy config %P', deployConfig);

      const searchConfig: SearchConfig = {};
      if (deployConfig.functionsFolderName) {
        searchConfig.functionsFolderNames = [deployConfig.functionsFolderName];
      }

      if (deployConfig.assetsFolderName) {
        searchConfig.assetsFolderNames = [deployConfig.assetsFolderName];
      }

      let { functions, assets } = await getListOfFunctionsAndAssets(
        deployConfig.cwd,
        searchConfig
      );

      if (deployConfig.noFunctions) {
        log('Disabling functions upload by emptying functions array');
        functions = [];
      }

      if (deployConfig.noAssets) {
        log('Disabling assets upload by emptying assets array');
        assets = [];
      }

      const config = {
        ...this.config,
        ...deployConfig,
        functions,
        assets,
      };

      return this.deployProject(config);
    } catch (err) {
      convertApiErrorsAndThrow(err);
    }
  }

  // request json without options
  async request(method: HTTPAlias, path: string): Promise<Response<unknown>>;
  // request json with options
  async request(
    method: HTTPAlias,
    path: string,
    options: OptionsOfJSONResponseBody
  ): Promise<Response<unknown>>;
  // request text
  async request(
    method: HTTPAlias,
    path: string,
    options: OptionsOfTextResponseBody
  ): Promise<Response<string>>;
  // general implementation
  async request(
    method: HTTPAlias,
    path: string,
    options: Options = {}
  ): Promise<unknown> {
    options.retry = {
      limit: this.config.retryLimit || RETRY_LIMIT,
      methods: ['GET', 'POST', 'DELETE'],
      statusCodes: [429],
    };
    return this.limit(() => this.client[method](path, options));
  }
}

export default TwilioServerlessApiClient;
